### DISCUSSION TITLE
ESPNow.recv(timeout)

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11112] [General]
I understand that "timeout" is in milliseconds, if left empty defaults to 5 minutes and if <0 does not return on timeout.
What is the maximum "timeout" can be set in milliseconds; is it 2**32ms?

MM

**Comment by glenn20** [comment]
I just checked the code as I was preparing my answer and realised that I think there is a regession in my code.

The maximum SHOULD be (2**31-1)ms (ie. 24 days), but I have been handling the negative value case incorrectly (I believe I fixed this some time ago - but noticed what I thought was my old code still in use).

From inspection of the **espnow-v1.19.1** and **espnow-g20** braches:

- the maximum safe value for timeout should be 2**31-1;
- negative values of timeout are being set to numbers over 2**31; and
- numbers larger than 2**32 will either:
  - throw an exception (I believe this is the case); or
  - be truncated to 32 bits (depending on behaviour of mp_obj_get_int()).

Thanks for prompting me to check this. I'll run some tests and update my code.

**Reply by mmliam** [reply]
Thanks Glenn, sorry for the late reply.

I don't intend on pushing the MAX; I was just curious what the MAX was as I couldn't find any documentation.

MM

**Reply by glenn20** [reply]
Ive pushed a [commit](https://github.com/micropython/micropython/pull/6515/commits/c944e74c18d8ec0a475b9562acf5132091728d1d) that changes the timeout behaviour:

- `timeout == 0`: Do not wait: return `(None, None)` if no messages available.
- `-2**31+1 <= timeout < 0`: Disable timeout: will wait forever for a message.
- `0 < timeout <= 2**31-1`: Set timeout to that number of milliseconds.
- `2**31 <= timeout || timeout <= -2**31`: Raise OverflowError.
